### PR TITLE
`NotebookSteps`, `steps` - migrate modifying queries to MLv2

### LINKED ISSUE #36516: [MLv2] [FE] `NotebookSteps`, `steps` - migrate modifying queries to MLv2
TODO:
- [x] remove `NotebookStep['update']`
- [x] update `NotebookStepUiComponentProps['updateQuery']`
- [x] remove `NotebookStepFn`
- [x] remove `NotebookStep['clean']`
- [x] Update `NotebookStepDef['valid']`
- [x] Update `NotebookStepDef['active']`
- [x] `NotebookStep['revert']` should return `Query`
- [x] `NotebookStep['previewQuery']` should be `Query`
- [ ] ~remove `legacyQuery` argument from~ - not yet, we need [Custom columns Milestone 1](https://github.com/metabase/metabase/issues/30513) first
  - [ ] ~`convertStageQueryToLegacyStageQuery`~
  - [ ] ~`getStageSteps`~
- [ ] ~Remove `Notebook['query']`~ - not yet, we need [Custom columns Milestone 1](https://github.com/metabase/metabase/issues/30513) first
- [x] `hasData` -> replace with `databaseId !== null` check
- [x] `Boolean(sourceQuery())` -> replace with `stageIndex > 0` check
- [x] `getQuestionSteps`
- [x] `steps.unit.spec.js`
  - [x] `getQuestionStepsForMBQLQuery`

### PR DESCRIPTION
**Pull request by kamilmielnik**
Closes #36516

### Description

~Blocked by #36690.~

~3 tests in `e2e-tests-joins-ee` fail because of #36690.~
2nd e2e test for #19745 is now also skipped, for [the same reason as the first one](https://metaboat.slack.com/archives/C0643FZ5E83/p1702365832126429?thread_ts=1702362649.548849&cid=C0643FZ5E83) ~- the tests should actually be removed in #36746~.

Don't miss the main `frontend/src/metabase/query_builder/components/notebook/lib/steps.ts` file which GitHub automatically hides the diff for.

### How to verify

GUI editor works. Especially adding/removing/previewing steps.

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx [hunk 1]
```diff
@@ -41,52 +43,42 @@ function NotebookSteps({
...
-
-      // mark the step as "closed" since we can assume
-      // it's been added or removed by the updateQuery
-      handleStepClose(step.id);
+    async (query: Query) => {
+      const updatedLegacyQuery = Lib.toLegacyQuery(query);
+      const updatedQuestion = question.setDatasetQuery(updatedLegacyQuery);
+      const updatedQuery = updatedQuestion.query() as StructuredQuery;
+      const cleanQuestion = updatedQuery.cleanNesting().question();
+      await updateQuestion(cleanQuestion);
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx:79]: Can this entire block be simply replaced with `question._setMLv2Query(query);`?

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.unit.spec.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
+import { createMockMetadata } from "__support__/metadata";
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/notebook/lib/steps.unit.spec.ts:1]: This file has a lot of changes and has been converted to TS - that's why it's shown as a new file.

File: frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx [hunk 3]
```diff
@@ -41,52 +43,42 @@ function NotebookSteps({
...
-      }
-
-      // mark the step as "closed" since we can assume
-      // it's been added or removed by the updateQuery
-      handleStepClose(step.id);
+    async (query: Query) => {
+      const updatedLegacyQuery = Lib.toLegacyQuery(query);
+      const updatedQuestion = question.setDatasetQuery(updatedLegacyQuery);
+      const updatedQuery = updatedQuestion.query() as StructuredQuery;
+      const cleanQuestion = updatedQuery.cleanNesting().question();
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx:78]: `dropStageIfEmpty` should do the same as `cleanNesting`

**Review by ranquild — COMMENTED** [review]
@kamilmielnik one comment about `cleanNesting`

File: frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx]: Do you really need to drop more than 1 stage at once? We discussed it with @snoe and I think it is never the case

Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx]: I'm not sure. `cleanNesting` worked recursively, that's why I did it like that.

Updated to try to drop only the updated stage: 88bfe40fa55f1d2c3f4fd050511dd80951680c09

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.ts [hunk 4]
```diff
@@ -1,173 +1,222 @@
...
+  const hasBreakouts = Lib.breakouts(query, stageIndex).length > 0;
+  const hasJoins = Lib.joins(query, stageIndex).length > 0;
+  const hasExpressions = Lib.expressions(query, stageIndex).length > 0;
+  const hasFilters = Lib.filters(query, stageIndex).length > 0;
+  const hasOrderBys = Lib.orderBys(query, stageIndex).length > 0;
+  const hasLimits = Lib.hasLimit(query, stageIndex);
+  const hasFields = Lib.fields(query, stageIndex).length > 0;
+
+  return (
+    hasJoins ||
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/notebook/lib/steps.ts:173]: it would be nice to have the same order as in `const` block

**Review by uladzimirdev — COMMENTED** [review]
I'll double check `frontend/src/metabase/query_builder/components/notebook/lib/steps.ts` later

File: frontend/src/metabase/query_builder/components/notebook/NotebookStepPreview.jsx [hunk 5]
```diff
@@ -38,9 +41,14 @@ class NotebookStepPreview extends Component {
 
   getPreviewQuestion(step) {
     const query = step.previewQuery;
-    const hasSuitableLimit = query.hasLimit() && query.limit() < 10;
+    const limit = Lib.currentLimit(query, -1);
+    const hasSuitableLimit = limit !== null && limit <= PREVIEW_ROWS_LIMIT;
+    const queryWithLimit = hasSuitableLimit
+      ? query
+      : Lib.limit(query, -1, PREVIEW_ROWS_LIMIT);
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/components/notebook/NotebookStepPreview.jsx]: Good catch, thanks! 5502015f31bb1be8446d52256867a61230498be2

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.ts (hunk 4 above)
Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/components/notebook/lib/steps.ts:173]: 9f8a79ad01777317e83a5c8f95817c3f88635488

**Review by ranquild — COMMENTED** [review]
Some parts are not migrated yet:

1. steps
https://github.com/metabase/metabase/blob/9f8a79ad01777317e83a5c8f95817c3f88635488/frontend/src/metabase/query_builder/components/notebook/lib/steps.ts#L194

`if (question.isStructured())` => ignore this, use `question._getMLv2Query()`
`legacyQuery = legacyQuery.cleanNesting();` => `dropStageIfEmpty`
`const stagedQueries = legacyQuery.queries()` => `stageCount`

2. Notebook

https://github.com/metabase/metabase/blob/aaf4a438168655ce5b374cf398bd86430baf78fb/frontend/src/metabase/query_builder/components/notebook/Notebook.tsx#L58

`query.clean` => `dropStageIfEmpty`

https://github.com/metabase/metabase/blob/aaf4a438168655ce5b374cf398bd86430baf78fb/frontend/src/metabase/query_builder/components/notebook/Notebook.tsx#L99

There is no `sourceTableOrCardId` atm, but it should be migrated eventually. I'll create a separate task

**Comment by kamilmielnik** [comment]
Some parts are not migrated yet:

> 1. steps

> `if (question.isStructured())` => ignore this, use `question._getMLv2Query()`

Done: 5386202fe431a90f41c3c19c2944fd6f3a5a38b


> `legacyQuery = legacyQuery.cleanNesting();` => `dropStageIfEmpty`

Done: a9aa4913cbd252aafe477d4302ab8461dc1cfb9

> `const stagedQueries = legacyQuery.queries()` => `stageCount`

The loop already uses `Lib.stageCount`. `legacyQuery.queries()` is still needed for backwards-compatibility and we won't be able to remove it without completing the Custom Columns epic.

> 2. Notebook

I didn't realize this component was in the scope of this task.

> `query.clean` => `dropStageIfEmpty`

Done: ceb7c12643b025d183efb77eafd943be22ad968

**Review by uladzimirdev — APPROVED** [review]

File: frontend/src/metabase-lib/query.ts [hunk 6]
```diff
@@ -43,6 +43,20 @@ export function dropStage(query: Query, stageIndex: number): Query {
...
+export function dropStageIfEmpty(query: Query, stageIndex: number): Query {
+  return ML.drop_stage_if_empty(query, stageIndex);
+}
+
+export function dropEmptyStages(query: Query): Query {
+  const stageIndexes = Array.from({ length: stageCount(query) })
+    .map((_, index) => index)
+    .reverse();
+
+  return stageIndexes.reduce((query, stageIndex) => {
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase-lib/query.ts]: I love it 9e0ec08033da4cdc44dad18dbd726ac146c35584

**Pull request merged**